### ISSUE TITLE
Shader language GLES2 type issue with mod

### ISSUE THREAD

**Issue by xix-xeaon** [godotengine/godot#35152]
**Godot version:**
Godot_v3.2-beta6_x11.64

**OS/device including version:**
Kubuntu 19.10

**Issue description:**
Running the code below (the second line) with GLES2 gives the following error in the output tab, the shader editor is unaware, and there some glitch drawing. Wrapping the expression in an int() works around it. It works in GLES3.
```
 drivers/gles2/shader_gles2.cpp:129 - SceneShaderGLES2: Fragment shader compilation failed:
0:1274(2): error: initializer of type float cannot be assigned to variable of type int

 drivers/gles2/shader_gles2.cpp:360 - Method/Function Failed, returning: __null
 drivers/gles2/shader_gles2.cpp:88 - Condition ' !version ' is true. returned: false
 drivers/gles2/shader_gles2.h:257 - Condition ' !version ' is true. returned: -1
 drivers/gles2/shader_gles2.h:257 - Condition ' !version ' is true. returned: -1
```

**Steps to reproduce:**
```
vec2 fragcord = UV * 10.0;
int val = int(fragcord.x) % 4;
```

**Comment by clayjohn** [comment]
This isn't a bug. That error report was proper. 

You can't take the modulo of a float. And you can't assign a float to an int without explicitly converting it to an int.

**Comment by xix-xeaon** [comment]
Not sure if we're understanding each other; int(fragcord.x) is an int, and 4 is an int, and the result of mod is an int. And it works perfectly well with GLES3.

**Comment by clayjohn** [comment]
I see, I misunderstood what you were saying. 

I reproduced the issue and was able to look into it further. GLSL 120 doesn't support the modulo operator (%), so internally it is converted to ``mod()``, but ``mod()`` takes two floats and returns a float. 

In the code that processes the replacement it should be adding the integer conversion as well.

**Comment by xix-xeaon** [comment]
Ah I see, that makes sense. Could there be float precision issue where returning float is 2.9999 and then becomes 2 instead of 3 after int?

**Comment by clayjohn** [comment]
Thats always a risk, I guess we could add "int(floor(0.5+mod(float(" instead.

**Comment by aaronfranke** [comment]
What is the bug exactly? A mod of two ints gives an int, a mod of two floats gives a float, the cast seems necessary and appropriate here. A mod accepting floats is not guaranteed to have a whole number as the output, so I don't see what kind of solution we could possibly implement. Adding `int(` and `)` as @clayjohn suggested sounds like it would cause % to always get rid of the decimals.

**Comment by clayjohn** [comment]
@aaronfranke this is for GLES2. It doesn't have ``vec2i``. It also doesn't technically support %. 

In GLES3 % is only for integer types. So in Godot, we restrict it to integer types. If you want the modulo of floats you use ``mod()``.

The issue is that in GLES2 we fake support for % by secretly converting it to ``mod(float(), float())`` in the shader compiler. But that means that it returns a floating point number, contrary to the spec for % which **always** takes and returns ints.

**Comment by aaronfranke** [comment]
> contrary to the spec for % which always takes and returns ints.

`%` works fine for floats in C# and Python, there's no reason we can't also make it work with floats. Why does the shader code have `float()` in there instead of just accepting the types as-is? Is it possible for us to both make a `mod(float(), float())` and an `int(mod(int(), int()))`

**Comment by clayjohn** [comment]
@aaronfranke We don't have control over GLSL. We do our best to follow the GLSL 330 spec which does not allow floats to be used with %. Also note C and C++ only allow ints to be used with % (GLSL is based on C).

**Comment by akien-mga** [comment]
> The issue is that in GLES2 we fake support for % by secretly converting it to mod(float(), float()) in the shader compiler. But that means that it returns a floating point number, contrary to the spec for % which always takes and returns ints.

Maybe we can drop that feature? Sounds confusing, especially if it differs from GLES3 shaders.

**Comment by clayjohn** [comment]
@akien-mga since GLES2 doesn't support % our options are to fake it with ``int(mod())`` or to ban it's use outright in GLES2.

**Comment by akien-mga** [comment]
We could also reimplement our own `_int_mod()` function to do it I guess? Good old integer division  and remainder. But otherwise casting the current mod result to int should do the trick indeed.

**Comment by aaronfranke** [comment]
Note that modulus and remainder are not the same thing. C, C++, C#, Java, and GDScript treat % as a remainder operator, and `fmod` in C++ is also a remainder operation. Python and Google Calculator treat % as a modulus operation, GLSL has `mod` as a modulus operation (at least according to [these docs](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/mod.xhtml), haven't tested it myself), and Godot has `posmod` and `fposmod`.

**Comment by akien-mga** [comment]
Fixed by #35202.
